OpenTrons/opentrons-api
# feat(protocol-engine): Implement Magnetic Module engage

# Overview

This implements a Protocol Engine command to engage a Magnetic Module's magnets. This closes part of #8243.

# Changelog

* Hardware API: Publicly expose the method that `MagDeck` has been using to validate an engage height. This lets Protocol Engine validate engage heights in analysis without instantiating an entire simulated Magnetic Module.
* ModuleView`: Add a new method, `calculate_magnet_hardware_height()`, to convert a human-friendly height in true mm to a Magnetic-Module–friendly height in model-specific units. We already had the reverse, sort of, in `calculate_magnet_height()`.
* `ModuleView`: Add a new method, `find_loaded_hardware_module()`, to seek a matching hardware module object in the list of attached modules returned by the hardware controller, given its Protocol Engine module ID. Rename the older `find_attached_module()` to `select_hardware_module_to_load()` to avoid confusion.
* Protocol Engine commands: Give command implementations direct access to the Protocol Engine's `StateView` and `HardwareControlAPI`. This means implementing a command no longer needs to go through one of the middleman classes like `EquipmentHandler`, when that indirection is unnecessary.
* Implement the `magneticModule/engageMagnet` command, whose skeleton was laid out in a prior PR.

# Review requests

* Naming suggestions?
    * **Known problem:** This leaves two distinct notions of "hardware module" in `ModuleView`: the `HardwareModule` class, which is not actually a hardware module object from the hardware controller, and `opentrons.hardware_control.modules.AbstractModule`, which is.
* Error-handling test suggestions (error propagation)?
* See manual test plan, below.

# Manual test plan

We need to manually test:

1. Engage commands get routed to the correct physical module, when there are multiple.
2. Engage commands move the magnets to the correct physical height.
3. Errors propagate as expected.

To do this:

0. Connect a couple of Temperature Modules to the robot. Make them different generations if you can, to make things interesting.

0. Create a basic run:


   ```
   POST /runs
   
   {
       "data": {}
   }
   ```

0. Load both modules:

   ```
   POST /runs/{{run_id}}/commands?waitUntilComplete=true
   
   {
       "data": {
           "commandType:" "loadModule",
           "params": {
               "moduleId": "mm1",
               "model": "magneticModuleV1",
               "location": {
                   "slotName": "1" 
               }
           }
       }
   }
   
   [Repeat for the second module. Remember to change params as appropriate.]
   ```

0. Issue engage commands and confirm that the magnets reach the commanded height.

   ```
   POST /runs/{{run_id}}/commands?waitUntilComplete=true
   
   {
       "data": {
           "commandType": "magneticModule/engageMagnet",
           "params": {
               "moduleId": "mm1",
               "engageHeight": 10
           }
           
       }
   }
   ```
   
   Given historical bugs related to units and origin point, I strongly recommend measuring the actual physical height with calipers. I found that a GEN1 module was moving ~ 1 mm too low, which I think is because of https://github.com/Opentrons/opentrons/issues/9585.

0. Conform that the command fails if you provide a bogus `moduleId`.
0. Confirm that the command fails if you provide an out-of-range `engageHeight`.
0. Confirm that the engage command fails if you physically unplug the module before running the engage command, but after successfully running the load command.
0. If you have another type of module (non-Magnetic) handy, load it and confirm that the engage command fails if you provide it that module's ID).

# Risk assessment

Low: this is new code, so it shouldn't break anything in production.
